### PR TITLE
Fix a bug with model indexes

### PR DESCRIPTION
**Pull request by kunichik**
Add serialization for a pair, fix a bug with model indexes

### CODE REVIEW

File: src/ipc/ipc-common/IPCProxyModelProperty.h [hunk 1]
```diff
@@ -139,16 +139,21 @@ class IPCProxyModelProperty : public facelift::ModelProperty<ModelDataType>
...
 
-                QList<ModelDataType> list;
-                m_proxy.ipc()->sendMethodCallWithReturn(requestMemberID, list, first, last);
+                std::pair<int, QList<ModelDataType>> listPair;
+                m_proxy.ipc()->sendMethodCallWithReturn(requestMemberID, listPair, first, last);
 
-                Q_ASSERT(list.size() == (last - first + 1));
+                auto &list = listPair.second;
+                first = listPair.first;// the server could change the number of returned items
+                if(list.empty()) {     // and even return an empty list
```
Comment (bitmouse) [inline src/ipc/ipc-common/IPCProxyModelProperty.h:147]: "or even return an empty list"

add space between 'if' and '('
add space between ; and //

File: src/ipc/ipc-common/IPCProxyModelProperty.h [hunk 2]
```diff
@@ -139,16 +139,21 @@ class IPCProxyModelProperty : public facelift::ModelProperty<ModelDataType>
...
+                    return retval;
+                }
+                last = first + list.size() - 1; 
 
                 for (int i = first; i <= last; ++i) {
                     m_cache.insert(i, list.at(i - first));
                 }
 
-                retval = list.at(row - first);
+                retval = list.at(row - first);//TODO use cache
```
Comment (bitmouse) [inline src/ipc/ipc-common/IPCProxyModelProperty.h:156]: add space between ; and //

File: src/ipc/ipc-common/IPCProxyModelProperty.h [hunk 3]
```diff
@@ -191,9 +196,11 @@ class IPCProxyModelProperty : public facelift::ModelProperty<ModelDataType>
             }
 
             if (first <= last) {
-                m_proxy.ipc()->sendAsyncMethodCall(requestMemberID, facelift::AsyncAnswer<QList<ModelDataType> >(&m_proxy, [this, first,
-                        last](QList<ModelDataType> list) {
+                m_proxy.ipc()->
+                sendAsyncMethodCall(requestMemberID, facelift::AsyncAnswer<std::pair<int, QList<ModelDataType>>>(&m_proxy,
```
Comment (bitmouse) [inline src/ipc/ipc-common/IPCProxyModelProperty.h:200]: fix formatting

File: src/ipc/ipc-common/ipc-common.h [hunk 4]
```diff
@@ -127,15 +127,15 @@ struct IPCTypeRegisterHandler<QList<Type> >
...
+struct IPCTypeRegisterHandler<QMap<QString, Type>>
 {
     typedef QMap<QString, typename IPCTypeRegisterHandler<Type>::SerializedType> SerializedType;
 
     template<typename OwnerType>
     static SerializedType convertToSerializedType(const QMap<QString, Type> &v, OwnerType &adapter)
     {
         SerializedType convertedValue;
-        for (const auto &key : v.keys()) {
+        for (const auto &key : v.keys()) {//TODO change to iterator for exclude search repeat (N*lnN->O(N))
```
Comment (bitmouse) [inline src/ipc/ipc-common/ipc-common.h:138]: add space before // or (preferably) move the comment above for loop

**Comment by kunichik** [comment]
> Please see [725e2ae](https://github.com/Pelagicore/facelift/commit/725e2aebe055c56c537d773194767a7342a980d3) for a proper fix

done, please verify and approve.

File: src/ipc/ipc-common/IPCProxyModelProperty.h [hunk 5]
```diff
@@ -139,10 +139,16 @@ class IPCProxyModelProperty : public facelift::ModelProperty<ModelDataType>
...
-                Q_ASSERT(list.size() == (last - first + 1));
+                std::tuple<int, QList<ModelDataType>> requestResult;
+                m_proxy.ipc()->sendMethodCallWithReturn(requestMemberID, requestResult, first, last);
+
+                first = std::get<0>(requestResult);
+                auto &list = std::get<1>(requestResult);
+                // The server could change the number of returned items or even return an empty list
+                if (list.empty()) {
+                    return retval;
+                }
```
Comment (kunichik) [inline src/ipc/ipc-common/IPCProxyModelProperty.h:150]: @jacky309, @bitmouse, here we had qt_assert for empty list returned by the server, so we return an empty Model data.

File: src/ipc/ipc-common/IPCProxyModelProperty.h [hunk 6]
```diff
@@ -139,10 +139,16 @@ class IPCProxyModelProperty : public facelift::ModelProperty<ModelDataType>
...
-                QList<ModelDataType> list;
-                m_proxy.ipc()->sendMethodCallWithReturn(requestMemberID, list, first, last);
-
-                Q_ASSERT(list.size() == (last - first + 1));
+                std::tuple<int, QList<ModelDataType>> requestResult;
+                m_proxy.ipc()->sendMethodCallWithReturn(requestMemberID, requestResult, first, last);
+
+                first = std::get<0>(requestResult);
+                auto &list = std::get<1>(requestResult);
+                // The server could change the number of returned items or even return an empty list
```
Comment (jacky309) [inline src/ipc/ipc-common/IPCProxyModelProperty.h:147]: this comment does not match the code below, where you only check if the list is empty. You need to check if you have received the requested item from the server.

File: src/ipc/ipc-common/IPCProxyModelProperty.h [hunk 7]
```diff
@@ -139,10 +139,16 @@ class IPCProxyModelProperty : public facelift::ModelProperty<ModelDataType>
...
+                if (list.empty()) {
+                    return retval;
```
Comment (jacky309) [inline src/ipc/ipc-common/IPCProxyModelProperty.h:149]: no early return path is needed here

Comment (kunichik) [reply src/ipc/ipc-common/IPCProxyModelProperty.h]: done, please look at the code.

File: src/ipc/ipc-common/IPCProxyModelProperty.h (hunk 6 above)
Comment (kunichik) [reply src/ipc/ipc-common/IPCProxyModelProperty.h]: done

File: src/ipc/ipc-common/IPCProxyModelProperty.h [hunk 8]
```diff
@@ -139,16 +135,19 @@ class IPCProxyModelProperty : public facelift::ModelProperty<ModelDataType>
...
 
-                QList<ModelDataType> list;
-                m_proxy.ipc()->sendMethodCallWithReturn(requestMemberID, list, first, last);
+                std::tuple<int, QList<ModelDataType>> requestResult;
+                m_proxy.ipc()->sendMethodCallWithReturn(requestMemberID, requestResult, first, last);
 
-                Q_ASSERT(list.size() == (last - first + 1));
+                first = std::get<0>(requestResult);
+                auto &list = std::get<1>(requestResult);
+                // Check if the server returned the requested item
```
Comment (jacky309) [inline src/ipc/ipc-common/IPCProxyModelProperty.h:143]: you are checking if the list is empty on the line below !

File: src/ipc/ipc-common/IPCProxyModelProperty.h [hunk 9]
```diff
@@ -139,16 +135,19 @@ class IPCProxyModelProperty : public facelift::ModelProperty<ModelDataType>
...
+                if (!list.isEmpty()) {
```
Comment (jacky309) [inline src/ipc/ipc-common/IPCProxyModelProperty.h:144]: Remove this useless condition. If the list is empty, no iteration will be done below anyway.

File: src/ipc/ipc-common/IPCProxyModelProperty.h [hunk 10]
```diff
@@ -168,7 +167,10 @@ class IPCProxyModelProperty : public facelift::ModelProperty<ModelDataType>
             requestItemsAsync(requestMemberID, previousIndex);
         }
 
-        return retval;
+        if (m_cache.exists(row)) {
```
Comment (jacky309) [inline src/ipc/ipc-common/IPCProxyModelProperty.h:170]: Get rid of multiple return paths with :
return (m_cache.exists(row) ? m_cache.get(row) : ModelDataType {});

Comment (bitmouse) [reply src/ipc/ipc-common/IPCProxyModelProperty.h:170]: pretty much no difference, IMO both versions look good

File: src/ipc/ipc-common/IPCProxyModelProperty.h (hunk 8 above)
Comment (kunichik) [reply src/ipc/ipc-common/IPCProxyModelProperty.h:143]: done - deleted comment at all.

File: src/ipc/ipc-common/IPCProxyModelProperty.h (hunk 9 above)
Comment (kunichik) [reply src/ipc/ipc-common/IPCProxyModelProperty.h:144]: done

File: src/ipc/ipc-common/IPCProxyModelProperty.h (hunk 10 above)
Comment (kunichik) [reply src/ipc/ipc-common/IPCProxyModelProperty.h:170]: ok, agree.
done

**Comment by kunichik** [comment]
all that was discussed fixed

**Pull request merged**